src/test/java/integration/TestRequestManager.java

```diff
@@ -0,0 +1,425 @@
+        testAddUserToProject(proj1.getProjectID(), proj1.getName(), user2ID, "admin", apiConstants);
+        testRemoveUserFromProject(proj1.getProjectID(), proj1.getName(), user2ID);
+        proj2 = testCreateProject(proj2.getName());
+        List<Long> projectIDs = new ArrayList<>();
+        projectIDs.add(proj1.getProjectID());
+        projectIDs.add(proj2.getProjectID());
+        testFetchAndSubscribeAll(projectIDs);
+        testUnsubscribeFromProject(proj1.getProjectID());
+        testFetchProjects();
+        testSubscribeToProject(proj1.getProjectID());
+        testAddUserToProject(proj2.getProjectID(), proj2.getName(), user2ID, "write", apiConstants);
+        testDeleteProject(proj1.getProjectID());
+        testLogout();
+        testLogin(user2ID, user2Pass);
+        testRemoveSelfFromProject(proj2.getProjectID());
+        testLogout();
+        testLogin(user1ID, user1Pass);
+        deleteProject(proj2.getProjectID());
+    }
+
+    private void testFetchPermissionConstants() throws InterruptedException {
+        logger.info("Fetching permission constants");
+        Semaphore waiter = new Semaphore(0);
+        PropertyChangeListener listener = (event) -> {
+            if (event.getPropertyName().equals(SessionStorage.PERMISSION_CONSTANTS)) {
+                apiConstants = (BiMap<String, Byte>) event.getNewValue();
+                waiter.release();
+            }
+        };
+        dataMgr.getSessionStorage().addPropertyChangeListener(listener);
+        reqMgr.fetchPermissionConstants();
+        if (!waiter.tryAcquire(5, TimeUnit.SECONDS)) {
+            Assert.fail("Acquire timed out");
+        }
+        dataMgr.getSessionStorage().removePropertyChangeListener(listener);
+    }
+
+    private void registerUser(String userID, String userFirstName,
+                              String userLastName, String userEmail, String userPass) throws InterruptedException, java.net.ConnectException {
+        logger.info(String.format("Registering user: %s", userID));
+        Semaphore waiter = new Semaphore(0);
+
+        req = new UserRegisterRequest(userID, userFirstName, userLastName, userEmail, userPass).getRequest(response -> {
+            // If registration fails, probably is already there.
+            waiter.release();
+        }, errHandler);
+        ws1.sendRequest(req);
+        if (!waiter.tryAcquire(5, TimeUnit.SECONDS)) {
+            Assert.fail("Acquire timed out");
+        }
+    }
+
+    private void testLogin(String userID, String userPass) throws InterruptedException {
+        logger.info("Logging in");
+        Semaphore waiter = new Semaphore(0);
+        final String[] senderToken = new String[1];
+        PropertyChangeListener listener = (event) -> {
+            if (event.getPropertyName().equals(SessionStorage.USERNAME)) {
+                if (userID.equals(event.getNewValue())) {
+                    waiter.release();
```

Comment:
We should unconditionally be releasing, and throwing Assert.fail() if there is an error. (or use Assert.AssertEquals(...))

That would allow us to identify what point the error is being generated from, instead of getting a cryptic "Acquire timed out"